sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala

```diff
@@ -114,8 +117,21 @@ case class BatchScanExec(
       // return an empty RDD with 1 partition if dynamic filtering removed the only split
       sparkContext.parallelize(Array.empty[InternalRow], 1)
     } else {
+      var finalPartitions = filteredPartitions
+
+      outputPartitioning match {
+        case p: KeyGroupedPartitioning =>
+          val partitionMapping = finalPartitions.map(s =>
+            s.head.asInstanceOf[HasPartitionKey].partitionKey() -> s).toMap
+          finalPartitions = p.partitionValuesOpt.get.map { partKey =>
```

Comment:
shall we move this logic to `lazy val filteredPartitions`?

Reply:
Hmm any obvious advantage of this? It looks the same to me 😅

Reply:
It's better to centralize the related code.

Reply:
I was hoping to re-use the logic code in #38950 too, which is not related to runtime filtering. If moving this close to the `if (oldRows.size < newRows.size)` check then I may have to duplicate it somewhere else?